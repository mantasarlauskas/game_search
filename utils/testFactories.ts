import { Category } from 'utils/types';

export function getCategories(): Category[] {
    return [
        {
            id: 18893,
            name: 'Feral Interactive',
            slug: 'feral-interactive',
            games_count: 107,
            image_background: 'https://media.rawg.io/media/games/157/15742f2f67eacff546738e1ab5c19d20.jpg',
            games: [
                {
                    slug: 'rise-of-the-tomb-raider',
                    name: 'Rise of the Tomb Raider',
                },
                {
                    slug: 'shadow-of-mordor',
                    name: 'Middle-earth: Shadow of Mordor',
                },
                {
                    slug: 'hitman',
                    name: 'Hitman',
                },
                {
                    slug: 'bioshock-2',
                    name: 'BioShock 2',
                },
                {
                    slug: 'bioshock-remastered',
                    name: 'BioShock Remastered',
                },
                {
                    slug: 'xcom-enemy-unknown',
                    name: 'XCOM: Enemy Unknown',
                },
            ],
        },
        {
            id: 405,
            name: 'Ubisoft',
            slug: 'ubisoft',
            games_count: 297,
            image_background: 'https://media.rawg.io/media/games/5e4/5e4bff02098b2b6f824c68286d5da1a6.jpg',
            games: [
                {
                    slug: 'far-cry-3',
                    name: 'Far Cry 3',
                },
                {
                    slug: 'for-honor',
                    name: 'For Honor',
                },
                {
                    slug: 'watch-dogs',
                    name: 'Watch Dogs',
                },
                {
                    slug: 'far-cry-5',
                    name: 'Far Cry 5',
                },
                {
                    slug: 'assassins-creed-origins',
                    name: 'Assassin\'s Creed Origins',
                },
                {
                    slug: 'rayman-legends',
                    name: 'Rayman Legends',
                },
            ],
        },
    ];
}
