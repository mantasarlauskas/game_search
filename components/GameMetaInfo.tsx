import Link from 'next/link';
import GameMetaBlock from 'components/GameMetaBlock';
import styles from 'styles/game-meta-info.module.scss';
import InfoList from 'components/InfoList';
import { Game } from 'utils/types';
import { Route } from 'utils/routes';

function GameMetaInfo({
    game: {
        platforms,
        metacritic,
        genres,
        released,
        developers,
        publishers,
        esrb_rating,
        tags,
        website,
    },
}: GameMetaInfoProps) {
    return (
        <div className={styles.root}>
            {!!platforms?.length && (
                <div className={styles.item}>
                    <GameMetaBlock title="Platforms">
                        <InfoList
                            list={platforms.map(({ platform }) => ({
                                ...platform,
                                url: `${Route.PLATFORMS}/${platform.id}`,
                            }))}
                        />
                    </GameMetaBlock>
                </div>
            )}
            {typeof metacritic === 'number' && (
                <div className={styles.item}>
                    <GameMetaBlock title="Metascore">
                        <div className={styles.score}>
                            {metacritic}
                        </div>
                    </GameMetaBlock>
                </div>
            )}
            {genres?.length > 0 && (
                <div className={styles.item}>
                    <GameMetaBlock title="Genres">
                        <InfoList
                            list={genres.map((genre) => ({
                                ...genre,
                                url: `${Route.GENRES}/${genre.id}`,
                            }))}
                        />
                    </GameMetaBlock>
                </div>
            )}
            {released && (
                <div className={styles.item}>
                    <GameMetaBlock title="Release date">
                        {released}
                    </GameMetaBlock>
                </div>
            )}
            {developers?.length > 0 && (
                <div className={styles.item}>
                    <GameMetaBlock title="Developers">
                        <InfoList list={developers} />
                    </GameMetaBlock>
                </div>
            )}
            {publishers?.length > 0 && (
                <div className={styles.item}>
                    <GameMetaBlock title="Publishers">
                        <InfoList list={publishers} />
                    </GameMetaBlock>
                </div>
            )}
            {tags?.length > 0 && (
                <div className={styles.item}>
                    <GameMetaBlock title="Tags">
                        <InfoList
                            list={tags.map((tag) => ({
                                ...tag,
                                url: `${Route.TAGS}/${tag.id}`,
                            }))}
                        />
                    </GameMetaBlock>
                </div>
            )}
            <div className={styles.item}>
                <GameMetaBlock title="Age rating">
                    {esrb_rating?.name || 'Not rated'}
                </GameMetaBlock>
            </div>
            {website && (
                <div className={styles.item}>
                    <GameMetaBlock title="Website">
                        <Link href={website} passHref>
                            {website}
                        </Link>
                    </GameMetaBlock>
                </div>
            )}
        </div>
    );
}

interface GameMetaInfoProps {
    game: Game;
}

export default GameMetaInfo;
