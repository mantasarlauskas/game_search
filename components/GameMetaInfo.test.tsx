import { render, screen } from '@testing-library/react';
import GameMetaInfo from 'components/GameMetaInfo';
import { getGames } from 'testing/testFactories';
import userEvent from '@testing-library/user-event';

describe('<GameMetaInfo />', () => {
    const [game] = getGames();

    it('matches snapshot', () => {
        const { container } = render(<GameMetaInfo game={game} />);
        expect(container).toMatchSnapshot();
    });

    it('renders meta info', () => {
        render(<GameMetaInfo game={game} />);
        expect(screen.getByText(/PC/)).toBeInTheDocument();
        expect(screen.getByText('Xbox One')).toBeInTheDocument();
        expect(screen.getByText('95')).toBeInTheDocument();
        expect(screen.getByText(/Shooter/)).toBeInTheDocument();
        expect(screen.getByText('Puzzle')).toBeInTheDocument();
        expect(screen.getByText('2011-04-18')).toBeInTheDocument();
        expect(screen.getByText('Singleplayer')).toBeInTheDocument();
        expect(screen.getByText('Everyone 10+')).toBeInTheDocument();
    });

    it('on click goes to correct page', () => {
        render(<GameMetaInfo game={game} />);
        userEvent.click(screen.getByText(/PC/));
        expect(window.location.href).toEqual('/platforms/4');
        userEvent.click(screen.getByText(/Shooter/));
        expect(window.location.href).toEqual('/genres/2');
        userEvent.click(screen.getByText('Singleplayer'));
        expect(window.location.href).toEqual('/tags/31');
    });
});
