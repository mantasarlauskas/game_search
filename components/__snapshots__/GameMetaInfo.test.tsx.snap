// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameMetaInfo /> matches snapshot 1`] = `
<div>
  <div
    class="root"
  >
    <div
      class="item"
    >
      <div
        class="root"
      >
        <div
          class="title"
        >
          Platforms
        </div>
        <div>
          <span
            class="item"
          >
            PC, 
          </span>
        </div>
        <div>
          <span
            class="item"
          >
            Xbox One
          </span>
        </div>
      </div>
    </div>
    <div
      class="item"
    >
      <div
        class="root"
      >
        <div
          class="title"
        >
          Metascore
        </div>
        <div
          class="score"
        >
          95
        </div>
      </div>
    </div>
    <div
      class="item"
    >
      <div
        class="root"
      >
        <div
          class="title"
        >
          Genres
        </div>
        <div>
          <span
            class="item"
          >
            Shooter, 
          </span>
        </div>
        <div>
          <span
            class="item"
          >
            Puzzle
          </span>
        </div>
      </div>
    </div>
    <div
      class="item"
    >
      <div
        class="root"
      >
        <div
          class="title"
        >
          Release date
        </div>
        2011-04-18
      </div>
    </div>
    <div
      class="item"
    >
      <div
        class="root"
      >
        <div
          class="title"
        >
          Tags
        </div>
        <div>
          <span
            class="item"
          >
            Singleplayer
          </span>
        </div>
      </div>
    </div>
    <div
      class="item"
    >
      <div
        class="root"
      >
        <div
          class="title"
        >
          Age rating
        </div>
        Everyone 10+
      </div>
    </div>
  </div>
</div>
`;
